docs/en/modules/tls.md

```diff
@@ -0,0 +1,291 @@
+# tls Module
+| Since  | Origin / Contributor  | Maintainer  | Source  |
+| :----- | :-------------------- | :---------- | :------ |
+| 2016-07-04 | [PhoeniX](https://github.com/djphoenix) | [PhoeniX](https://github.com/djphoenix) | [tls.c](../../../app/modules/tls.c)|
+
+**SSL/TLS support**
+
+Secure connections use **TLS 1.1** with the following cipher suites:
+
+- `TLS_RSA_WITH_AES_128_CBC_SHA`
+- `TLS_RSA_WITH_AES_256_CBC_SHA`
+- `TLS_RSA_WITH_RC4_128_SHA`
+- `TLS_RSA_WITH_RC4_128_MD5`
+
+This specification is imposed by the [axTLS library](http://axtls.sourceforge.net/specifications.htm) used by the SDK. 
+
+
+This controls certificate verification when SSL is in use.
+
+## tls.createConnection()
+
+Creates TLS connection.
+
+#### Syntax
+`tls.createConnection()`
+
+#### Parameters
+none
+
+#### Returns
+tls.socket sub module
+
+#### Example
+
+```lua
+tls.createConnection()
+```
+
+# tls.socket Module
+
+## tls.socket:close()
+
+Closes socket.
+
+#### Syntax
+`close()`
+
+#### Parameters
+none
+
+#### Returns
+`nil`
+
+#### See also
+[`tls.createConnection()`](#tlscreateconnection)
+
+## tls.socket:connect()
+
+Connect to a remote server.
+
+#### Syntax
+`connect(port, ip|domain)`
+
+#### Parameters
+- `port` port number
+- `ip` IP address or domain name string
+
+#### Returns
+`nil`
+
+#### See also
+[`tls.socket:on()`](#tlssocketon)
+
+## tls.socket:dns()
+
+Provides DNS resolution for a hostname.
+
+#### Syntax
+`dns(domain, function(tls.socket, ip))`
+
+#### Parameters
+- `domain` domain name
+- `function(tls.socket, ip)` callback function. The first parameter is the socket, the second parameter is the IP address as a string.
+
+#### Returns
+`nil`
+
+#### Example
+```lua
+sk = tls.createConnection()
+sk:dns("google.com", function(conn, ip) print(ip) end)
+sk = nil
+```
+
+## tls.socket:getpeer()
+
+Retrieve port and ip of peer.
+
+#### Syntax
+`getpeer()`
+
+#### Parameters
+none
+
+#### Returns
+- `ip` of peer
+- `port` of peer
+
+## tls.socket:hold()
+
+Throttle data reception by placing a request to block the TCP receive function. This request is not effective immediately, Espressif recommends to call it while reserving 5*1460 bytes of memory.
+
+#### Syntax
+`hold()`
+
+#### Parameters
+none
+
+#### Returns
+`nil`
+
+#### See also
+[`tls.socket:unhold()`](#tlssocketunhold)
+
+## tls.socket:on()
+
+Register callback functions for specific events.
+
+#### Syntax
+`on(event, function())`
+
+#### Parameters
+- `event` string, which can be "connection", "reconnection", "disconnection", "receive" or "sent"
+- `function(tls.socket[, string])` callback function. The first parameter is the socket.
+If event is "receive", the second parameter is the received data as string.
+If event is "reconnection", the second parameter is the reason of connection error (string).
+
+#### Returns
+`nil`
+
+#### Example
+```lua
+srv = tls.createConnection()
+srv:on("receive", function(sck, c) print(c) end)
+srv:connect(443,"google.com")
```

Comment:
The `:connect` should be moved *after* the `:on` registration to avoid a potential race - when pasting in the example into the console, the connection may be established before the callback registration has happened. I discovered this the hard way (a whole bunch of `printf`s later).

Comment also applies to the `net.socket:on()` entry.